src/libs/actions/OnyxUpdateManager.ts

```diff
@@ -66,32 +70,101 @@ export default () => {
                     lastUpdateIDFromServer,
                     previousUpdateIDFromServer,
                     lastUpdateIDAppliedToClient,
                 });
+
+                // Get the missing Onyx updates from the server
                 canUnpauseQueuePromise = App.getMissingOnyxUpdates(lastUpdateIDAppliedToClient, previousUpdateIDFromServer);
             }
 
-            canUnpauseQueuePromise.finally(() => {
-                OnyxUpdates.apply(updateParams).finally(() => {
+            // This function will apply the deferred updates in order after the missing updates are fetched
+            function applyDeferredUpdates() {
+                // If lastUpdateIDAppliedToClient is null, then no updates were applied,
+                // therefore something must have gone wrong and we should handle the problem.
+                if (!lastUpdateIDAppliedToClient) {
+                    // TODO: Handle this case
+                    return;
+                }
+
+                function unpauseQueueAndReset() {
                     console.debug('[OnyxUpdateManager] Done applying all updates');
                     Onyx.set(ONYXKEYS.ONYX_UPDATES_FROM_SERVER, null);
                     SequentialQueue.unpause();
+                }
+
+                // If "updateAfterReconnectApp" is set, it means that case (1) was triggered and we only need to apply the one update, not the deferred ones.
+                if (deferredUpdateBeforeReconnect) {
+                    OnyxUpdates.apply(deferredUpdateBeforeReconnect).finally(unpauseQueueAndReset);
+                    deferredUpdateBeforeReconnect = undefined;
+                    return;
+                }
+
+                // If case (2) was triggered, then we need to sort the deferred updates by lastUpdateID and apply them in order.
+                const sortedDeferredUpdates = deferredUpdates.sort((a, b) => {
+                    const aLastUpdateID = Number(a.lastUpdateID) || 0;
+                    const bLastUpdateID = Number(b.lastUpdateID) || 0;
+
+                    return aLastUpdateID - bLastUpdateID;
                 });
-            });
+
+                // In order for the deferred updates to be applied correctly in order,
+                // we need to check if there are any gaps deferred updates.
+                function validateUpdatesAreChained(): boolean {
+                    for (let i = 0; i < sortedDeferredUpdates.length - 1; i++) {
+                        const reversedDeferredUpdates = sortedDeferredUpdates.toReversed();
+                        const currentUpdate = reversedDeferredUpdates[i];
+                        const previousUpdate = reversedDeferredUpdates[i + 1];
+
+                        if (Number(previousUpdate.lastUpdateID) !== Number(currentUpdate.previousUpdateID)) {
+                            return false;
+                        }
+                    }
```

Comment:
This should work since the keys will be ordered: 
![image](https://github.com/Expensify/App/assets/5818999/94f2b150-3ebb-4f18-a00a-b2d2b4626814)


```suggestion
                    let isFirst = true;
                    let previousValidLastUpdateID = 0;
                    for (const [key, value] of Object.entries(deferredUpdates)) {
                        // If it's the first one, we need to skip it since we won't have the previousUpdateID
                        // if the previousUpdateID key exists, then we can just move on since there's no gap
                        if (isFirst || deferredUpdates[value.previousUpdateID]) {
                            previousValidLastUpdateID = value.lastUpdateID;
                            isFirst = false;
                            continue;
                        }
                        return false;
                    }
```